crates/ruff_python_parser/src/string_token_flags.rs

```diff
@@ -0,0 +1,295 @@
+    /// but can have no other prefixes.
+    Format,
+
+    /// The string has a `b` or `B` prefix.
+    /// This means that the string is a sequence of `int`s at runtime,
+    /// rather than a sequence of `str`s.
+    /// Bytestrings can also be raw strings,
+    /// but can have no other prefixes.
+    Bytes,
+
+    /// A string that has has any one of the prefixes
+    /// `{"rf", "rF", "Rf", "RF", "fr", "fR", "Fr", "FR"}`
+    /// Semantically, these all have the same meaning:
+    /// the string is both an f-string and a raw-string
+    RawFormat,
+
+    /// A string that has has any one of the prefixes
+    /// `{"rb", "rB", "Rb", "RB", "br", "bR", "Br", "BR"}`
+    /// Semantically, these all have the same meaning:
+    /// the string is both an bytestring and a raw-string
+    RawBytes,
+}
+
+impl TryFrom<char> for StringPrefix {
+    type Error = String;
+
+    fn try_from(value: char) -> Result<Self, String> {
+        let result = match value {
+            'r' | 'R' => Self::Raw,
+            'u' | 'U' => Self::Unicode,
+            'b' | 'B' => Self::Bytes,
+            'f' | 'F' => Self::Format,
+            _ => return Err(format!("Unexpected prefix '{value}'")),
+        };
+        Ok(result)
+    }
+}
+
+impl TryFrom<[char; 2]> for StringPrefix {
+    type Error = String;
+
+    fn try_from(value: [char; 2]) -> Result<Self, String> {
+        match value {
+            ['r' | 'R', 'f' | 'F'] | ['f' | 'F', 'r' | 'R'] => Ok(Self::RawFormat),
+            ['r' | 'R', 'b' | 'B'] | ['b' | 'B', 'r' | 'R'] => Ok(Self::RawBytes),
+            _ => Err(format!("Unexpected prefix '{}{}'", value[0], value[1])),
+        }
+    }
+}
+
+impl StringPrefix {
+    const fn as_flags(self) -> StringFlags {
+        match self {
+            Self::Bytes => StringFlags::B_PREFIX,
+            Self::Format => StringFlags::F_PREFIX,
+            Self::Raw => StringFlags::R_PREFIX,
+            Self::RawBytes => StringFlags::R_PREFIX.union(StringFlags::B_PREFIX),
+            Self::RawFormat => StringFlags::R_PREFIX.union(StringFlags::F_PREFIX),
+            Self::Unicode => StringFlags::U_PREFIX,
+        }
+    }
+}
+
+#[derive(Default, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct StringKind(StringFlags);
+
+impl StringKind {
+    pub(crate) const fn from_prefix(prefix: StringPrefix) -> Self {
+        Self(prefix.as_flags())
+    }
+
+    /// Does the string have a `u` or `U` prefix?
+    pub const fn is_ustring(self) -> bool {
+        self.0.contains(StringFlags::U_PREFIX)
+    }
+
+    /// Does the string have an `r` or `R` prefix?
+    pub const fn is_rawstring(self) -> bool {
+        self.0.contains(StringFlags::R_PREFIX)
+    }
+
+    /// Does the string have an `f` or `F` prefix?
+    pub const fn is_fstring(self) -> bool {
+        self.0.contains(StringFlags::F_PREFIX)
+    }
+
+    /// Does the string have a `b` or `B` prefix?
+    pub const fn is_bytestring(self) -> bool {
+        self.0.contains(StringFlags::B_PREFIX)
+    }
```

Comment:
nit: I'd prefer to have the `string` prefix separate for readability

```suggestion
    /// Does the string have a `u` or `U` prefix?
    pub const fn is_u_string(self) -> bool {
        self.0.contains(StringFlags::U_PREFIX)
    }

    /// Does the string have an `r` or `R` prefix?
    pub const fn is_raw_string(self) -> bool {
        self.0.contains(StringFlags::R_PREFIX)
    }

    /// Does the string have an `f` or `F` prefix?
    pub const fn is_f_string(self) -> bool {
        self.0.contains(StringFlags::F_PREFIX)
    }

    /// Does the string have a `b` or `B` prefix?
    pub const fn is_byte_string(self) -> bool {
        self.0.contains(StringFlags::B_PREFIX)
    }
```